### PR TITLE
Added libpcap-based backend

### PR DESCRIPTION
**Pull request by moosingin3space**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -45,6 +45,10 @@ version = "0.2.*"
 version = "0.42.*"
 optional = true
 
+[dependencies.pcap]
```
Comment (mrmonday) [inline Cargo.toml]: Do you not also need to add `pcap = ["pcap"]` to the features section?

File: src/datalink/libpcap.rs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+/// Get a list of available network interfaces for the current machine.
+pub fn interfaces() -> Vec<NetworkInterface> {
+    if let Ok(devices) = pcap::Device::list() {
+        devices.iter().enumerate().map(|(i, dev)| {
+            NetworkInterface {
+                name: dev.name.clone(),
+                index: i as u32,
+                mac: None,
+                ips: None,
```
Comment (mrmonday) [inline src/datalink/libpcap.rs]: Does `pcap` not expose IP addresses or MAC addresses for devices?

File: src/datalink/mod.rs [hunk 3]
```diff
@@ -54,8 +54,12 @@ mod backend;
 #[cfg(feature = "netmap")]
 pub mod netmap;
 
+#[cfg(feature = "pcap")]
+pub mod libpcap;
```
Comment (mrmonday) [inline src/datalink/mod.rs]: I'd rather this be called `pcap`, unless there's a particular reason not to?

**Comment by mrmonday** [comment]
Looks good overall, thanks for doing this! 🙂

File: Cargo.toml (hunk 1 above)
Comment (moosingin3space) [inline Cargo.toml]: No, every optional dependency creates a feature with the same name.

File: src/datalink/libpcap.rs (hunk 2 above)
Comment (moosingin3space) [inline src/datalink/libpcap.rs]: It does not.

File: Cargo.toml (hunk 1 above)
Comment (mrmonday) [inline Cargo.toml]: Oh cool, is that a new thing or has Cargo always done that?

File: src/datalink/libpcap.rs (hunk 2 above)
Comment (mrmonday) [inline src/datalink/libpcap.rs]: That's a shame.

File: Cargo.toml (hunk 1 above)
Comment (moosingin3space) [inline Cargo.toml]: I don't know, but @alexcrichton told me about it las tweek.

File: src/datalink/mod.rs (hunk 3 above)
Comment (moosingin3space) [inline src/datalink/mod.rs]: The module or the feature?

Comment (mrmonday) [inline src/datalink/mod.rs]: The module. I'm actually happy either way, as long as it's consistent

File: src/datalink/libpcap.rs (hunk 2 above)
Comment (moosingin3space) [inline src/datalink/libpcap.rs]: They have an open issue for it, might be worth helping them out: https://github.com/ebfull/pcap/issues/13

Comment (moosingin3space) [inline src/datalink/libpcap.rs]: Although reading libpcap documentation suggests that MAC addresses aren't visible in their API.

**Comment by mrmonday** [comment]
Great stuff, thank you!

@homu r+

**Pull request merged**